### ISSUE TITLE
Sample application, memory leaks found by LeakCanary

### ISSUE THREAD

**Issue by mat8854** [osmdroid/osmdroid#1463]
## Issue Type
[ X] Bug

## Description and/or steps/code to reproduce the problem

I was using the demo application and noticed that `LinearLayout` is leaked when calling a sample and moving back to the ListView. I added screenshots from three different cases below.

### If it's a bug, version(s) of android this affects:

All

### Version of osmdroid the issue relates to:

6.1.5

![Screenshot_20191227-105837](https://user-images.githubusercontent.com/2643309/71513056-1a071300-2899-11ea-9e80-a10efd57c810.png)
![Screenshot_20191227-105847](https://user-images.githubusercontent.com/2643309/71513057-1a071300-2899-11ea-8640-e33c5eded3fe.png)
![Screenshot_20191227-105909](https://user-images.githubusercontent.com/2643309/71513058-1a9fa980-2899-11ea-91c3-aa93328ef4d4.png)

**Comment by monsieurtanuki** [comment]
@mat8854 You're right, and it's a pain in the neck when you have a smartphone with little space left because each time it creates some kind of 50Mb core dump.

**Comment by mat8854** [comment]
@monsieurtanuki The sample application still uses version `1.6.x`. It can be [upgraded](https://square.github.io/leakcanary/upgrading-to-leakcanary-2.0/) to version `2`.

For the leak issue [WeakReference](https://developer.android.com/reference/java/lang/ref/WeakReference) is an option. Did not use that construct before.

More here: https://medium.com/google-developer-experts/finally-understanding-how-references-work-in-android-and-java-26a0d9c92f83

**Comment by spyhunter99** [comment]
just am fyi, the version published to the paystore has this feature disabled

**Comment by mat8854** [comment]
I cross checked it with the old `LeakCanary 1.6`. When changing `expListView` into a `WeakReference` and adjusting the three access calls the leaks seem to be fixed.
```
WeakReference<ExpandableListView> expListView;
...
expListView = new WeakReference<>((ExpandableListView) root.findViewById(R.id.lvExp));
...
expListView.get().setOnChildClickListener(new ExpandableListView.OnChildClickListener() {
...
expListView.get().setAdapter(listAdapter);
```
File: [SamplesMenuFragment](https://github.com/osmdroid/osmdroid/blob/master/OpenStreetMapViewer/src/main/java/org/osmdroid/samplefragments/ui/SamplesMenuFragment.java).

**Comment by monsieurtanuki** [comment]
@mat8854 Let's try this `WeakReference` fix! Please send a PR when you're ready.

**Comment by mat8854** [comment]
@monsieurtanuki Yes okay, I am not 100% sure why LeakCanary `1.6` and `2` behave different. Most likely using a `WeakReference` will not break anything. I push the changed code onto another PR because this #ID is already merged.

**Comment by monsieurtanuki** [comment]
@mat8854 In those cases I name the new branch something like "bug/#1463_2". I don't know if there's a best practice.

**Comment by monsieurtanuki** [comment]
@mat8854 I've just tried the combination of leakcanary 2.0 and WeakReference.
It's only a bit better than before.
I still sometimes have a canary showing his face when I enter a new demo (which is fishy, especially for a bird) (with a message like "1 retained objects, tap to dump heap / App visible, wait until 5 retained"), but most of the times when I exit the demo within say 5 seconds it's OK, I have a message saying "All retained objects were garbage collected".

In short, it looks like there's still a leak, and WeakReference (or leakcanary 2.0) makes it a bit less painful.

**Comment by mat8854** [comment]
@monsieurtanuki I tested on a real hardware device (Pixel 3a) back and forth (20+ times ) between fragments and no leaks were found. Testing on a device with lower memory may show leaks earlier.

**Comment by monsieurtanuki** [comment]
@mat8854 I always test on a real hardware device (Samsung Galaxy Core Prime) (which happens to be my smartphone), with low memory. And it leaks almost always when playing a "More Samples" demo: my smartphone is a real canary ;)

In short:
* there were leaks, according to leakcanary 1.6
* with your changes (WeakReference + leakcanary 2.0) the leaks are now invisible on your device
* the leaks seem to be still there, but now maybe only on low memory smartphones

There's a good chance that the leaks are related to the fragment management inside the demo app, and not the osmdroid core.
Now it would be my turn to fix things (as you don't experience leaks anymore and I still do), but I don't believe in fragments.
We need a fragment+leaks expert.

**Comment by mat8854** [comment]
@monsieurtanuki Yes, the code in `SamplesMenuFragment` is not easy to read. There are already many comments about leaks, saved instances, stacks, a commented `System.gc()` call ... The whole Fragement handling looks "fragile".

I am pretty sure at least one leak is related to a lifecycle issue in `SamplesMenuFragment`.

**Comment by monsieurtanuki** [comment]
@mat8854 I've just pulled the latest version of the code, and the good surprise is that there are no leak anymore even on my canary smartphone. Good job! Feel free to close the issue.

**Comment by monsieurtanuki** [comment]
Memory leaks still happen once in a while, even with the `WeakReference` fix.

According to https://square.github.io/leakcanary/fundamentals/:

> Most memory leaks are caused by bugs related to the lifecycle of objects. Here are a few common Android mistakes:
o Adding a Fragment to the backstack without clearing its view fields in Fragment.onDestroyView() (see this [great StackOverflow answer](https://stackoverflow.com/questions/59503689/could-navigation-arch-component-create-a-false-positive-memory-leak/59504797#59504797))

According to "a great StackOverflow answer":
> You have to null out your references to the views in onDestroyView as that's the sign that the view is no longer being used by the Fragment system and it can be safely garbage collected

I'll have a look at it on my canary smartphone.
@mat8854 The currently merged fix is still better than the previous version.

**Comment by monsieurtanuki** [comment]
So far so good. The solution seems to be:
* in all `Fragment`s, the `View`s must be `null`ed in `onDestroyView` (which is already the case in [`BaseSampleFragment`](https://github.com/osmdroid/osmdroid/blob/master/OpenStreetMapViewer/src/main/java/org/osmdroid/samplefragments/BaseSampleFragment.java#L120) because of an old memory leak issue #335)
* a good practice seems to call the `super` statement _last_ in `onDestroyView`

As there are about 140 `Fragment`s in the demo app, the fix is not going to be instant (though it's easy).
I'm working on it.

**Comment by monsieurtanuki** [comment]
Life is short: I limited the fix to the main `Fragment`s.
If we find fragments related memory leaks, now we now how to deal with them.
Fixed by #1487.
